hazelcast-sql/src/main/java/com/hazelcast/sql/impl/type/QueryDataType.java

```diff
@@ -272,56 +255,47 @@ private void collectNestedTypes(final QueryDataType dataType, final Map<String,
 
     @Override
     public void readData(ObjectDataInput in) throws IOException {
-        this.converter = Converters.getConverter(in.readInt());
-        if (!converter.getTypeFamily().equals(QueryDataTypeFamily.OBJECT)) {
+        converter = Converters.getConverter(in.readInt());
+        if (converter.getTypeFamily() != QueryDataTypeFamily.OBJECT) {
             return;
         }
 
-        this.objectTypeKind = in.readInt();
-        this.objectTypeName = in.readString();
-        this.objectTypeMetadata = in.readString();
+        objectTypeKind = TypeKind.of(in.readInt());
+        objectTypeName = in.readString();
+        objectTypeMetadata = in.readString();
+
         if (!isCustomType()) {
             return;
         }
 
-        this.objectFields = new ArrayList<>();
-        final int typeMapSize = in.readInt();
-        final Map<String, QueryDataType> nestedTypes = new HashMap<>();
+        objectFields = new ArrayList<>();
+        int typeMapSize = in.readInt();
+        Map<String, QueryDataType> nestedTypes = new HashMap<>();
         nestedTypes.put(objectTypeName, this);
 
         for (int i = 0; i < typeMapSize; i++) {
-            final Integer currentTypeKind = in.readInt();
-            final String currentTypeName = in.readString();
-            final String currentTypeMetadata = in.readString();
-
-            final int fieldsSize = in.readInt();
-            final QueryDataType currentType = nestedTypes.computeIfAbsent(currentTypeName, s -> {
-                // TODO: simplify?
-                final QueryDataType fieldType = new QueryDataType(currentTypeName);
-                fieldType.setObjectTypeMetadata(currentTypeMetadata);
-                fieldType.setObjectTypeKind(currentTypeKind);
-                return fieldType;
-            });
+            TypeKind currentTypeKind = TypeKind.of(in.readInt());
+            String currentTypeName = in.readString();
+            String currentTypeMetadata = in.readString();
+
+            int fieldsSize = in.readInt();
+            QueryDataType currentType = nestedTypes.computeIfAbsent(currentTypeName,
+                    n -> new QueryDataType(currentTypeName, currentTypeKind, currentTypeMetadata));
 
             for (int j = 0; j < fieldsSize; j++) {
-                final String fieldName = in.readString();
-                final int fieldConverterId = in.readInt();
-                final int fieldTypeKind = in.readInt();
-                final String fieldTypeName = in.readString();
-                final String fieldTypeMetadata = in.readString();
+                String fieldName = in.readString();
```

Comment:
I do not fully understand this. Why do we support only 2 levels of nesting of QueryDataTypes and do it by code copy-paste instead of recursion?

Reply:
I have refactored `readData()` to make it symmetric to `writeData()`. The final version doesn't contain any duplicate code. `writeData()` collects all types into a map beforehand to avoid infinite recursion. Then, it writes each type with its direct descendants, i.e. each subtree, in an arbitrary order. `readData()` first puts the root into the type map. Then, it reads all subtrees by creating a type only if it is not created before using the type map. Even though subtrees don't lie in a particular order, the children of all subtrees will eventually be populated, including the root. Writing distinct subtrees and reading them out-of-order with eventual consistency (linking) is a very clever idea. Not easily readable, but quite efficient and elegant.

Reply:
current implementation makes the classes more coupled as it does not use `writeData` for `QueryDataTypeField` even though it is IDS. It is also not fully consistent as top level QDT is serialized in different form than QDTs nested in it. But that should be acceptable.

Another consequence of current implementation (but it might have been the same before) is that if a mapping has many fields referencing the same UDT (and they contains nested UDT fields), UDT definition will be repeated for each field, as we invoke `collectCustomTypes` on QDT, not mapping level.

Reply:
> Writing distinct subtrees and reading them out-of-order with eventual consistency (linking) is a very clever idea. Not easily readable, but quite efficient and elegant.

Exactly. I wondered if it is possible to do that in explicit order (eg. starting from leaves) but that would not work for cyclic type definitions.

Reply:
IMO it is worth adding the above description as a comment in the code

Reply:
but generally the implementation is ok

Reply:
9a0e8973